extensions/amp-skimlinks/0.1/affiliate-link-resolver.js

```diff
@@ -0,0 +1,227 @@
+
+import {parseUrlDeprecated} from '../../../src/url';
+
+import {DOMAIN_RESOLVER_API_URL} from './constants';
+import {createAnchorReplacementTuple, createTwoStepsResponse} from '../../../src/service/link-rewrite/link-rewrite-helpers';
+
+// Can be monetized
+export const STATUS__AFFILIATE = 'affiliate';
+// Can't be monetized but can be tracked
+export const STATUS__NON_AFFILIATE = 'non-affiliate';
+// Can't be monetized and can't be tracked
+export const STATUS__IGNORE_LINK = 'ignore';
+// We don't know yet if the link can me monitized. Will be considered as
+// 'can be monetized' until we have the answer from the API.
+export const STATUS__UNKNOWN = 'unknown';
+
+export default class AffiliateLinkResolver {
+  /**
+   *
+   * @param {*} xhr
+   * @param {*} skimOptions
+   * @param {*} waypoint
+   */
+  constructor(xhr, skimOptions, waypoint) {
+    this.xhr_ = xhr;
+    this.skimOptions_ = skimOptions;
+    this.waypoint_ = waypoint;
+    this.domains_ = {};
+    // Promise of the first request to beacon so we can
+    // access API data outside of the linkRewriter/LinkResolver flow.
+    this.firstApiRequest = null;
+  }
+
+  /**
+   *
+   * @param {*} anchorList
+   * @return {Promise}
+   */
+  resolveUnknownAnchors(anchorList) {
+    const alreadyResolvedTupleList = this.mapToAnchorReplacementTuple_(anchorList);
+    let willBeResolvedPromise = null;
+
+    const domainsToAsk = this.getNewDomains_(anchorList);
+    if (domainsToAsk.length) {
+      // Set domains to STATUS__UNKNOWN to mark them as already requested.
+      this.markDomainsAsUnknown(domainsToAsk);
+      // Get anchors waiting for the API response to be resolved.
+      const unknownAnchors = this.getUnknownAnchors_(anchorList, domainsToAsk);
+      willBeResolvedPromise = this.resolvedUnknownAnchorsAsync_(unknownAnchors,
+          domainsToAsk);
+    }
+
+    // Returns an object with a sync reponse and an async response.
+    return new createTwoStepsResponse(alreadyResolvedTupleList, willBeResolvedPromise);
+  }
+
+
+  /**
+   * Build list of AnchorReplacementTuple
+   * @param {*} anchorList
+   */
+  mapToAnchorReplacementTuple_(anchorList) {
```

Comment:
Random Q: Does the hanging `_` mean that it's a private method ? If so, do they follow any convention on it's order in a class (Either all private first or last) ?

There doesn't seem to be any particular order here which I guess answers my question 🙈

Reply:
Yes _ means it's private, I have added more jsdoc about it in an other PR. I also moved functions around to put the public methods at the top but it's my own convention I don't think they have a rule.